[AST] Member lookup in the GSB needs to understand equivalence classes with concrete types.

huonw opened the pull request:
A constraint like `Parameter == SomethingConcrete` means references to
`Parameter` in that context behave like `SomethingConcrete`, including for name
resolution. Handling the parameter as just a parameter type means that it won't
find any non-protocol nested types (i.e. things other than associated types and
protocol typealiases are invisible).

Fixes rdar://problem/42136457 and fixes #50772.

huonw commented:
@swift-ci please smoke test

slavapestov reviewed (commented):
I’m not sure this is the right fix. Elsewhere we can deal with interface types that canonicalize down to concrete types, and mapping them to concrete types here loses the spelling.

huonw committed:
[AST] Member lookup in the GSB needs to understand equivalence classes with concrete types.

A constraint like `Parameter == SomethingConcrete` means references to
`Parameter` in that context behave like `SomethingConcrete`, including for name
resolution. Handling the parameter as just a parameter type means that it won't
find any non-protocol nested types (i.e. things other than associated types and
protocol typealiases are invisible).

Fixes rdar://problem/42136457 and SR-8240.

huonw commented:
That's fair. I've now changed it so the GSB can do this sort of concrete type lookup in a similar manner to how it does a superclass lookup.

slavapestov commented:
Thank you! @DougGregor should still take a look, but this seems like a more reasonable fix.

DougGregor reviewed (approved):
LGTM
